Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp

```diff
@@ -26,16 +26,69 @@
 #include "config.h"
 #include "WakeLock.h"
 
+#include "Document.h"
+#include "EventLoop.h"
 #include "Exception.h"
+#include "FeaturePolicy.h"
 #include "JSDOMPromiseDeferred.h"
+#include "JSWakeLockSentinel.h"
+#include "PermissionController.h"
+#include "PermissionQuerySource.h"
+#include "PermissionState.h"
+#include "WakeLockManager.h"
+#include "WakeLockSentinel.h"
 
 namespace WebCore {
 
-WakeLock::WakeLock() = default;
+WakeLock::WakeLock(Document* document)
+    : ContextDestructionObserver(document)
+{
+}
+
+// https://www.w3.org/TR/screen-wake-lock/#the-request-method
+void WakeLock::request(WakeLockType lockType, Ref<DeferredPromise>&& promise)
+{
+    auto* document = this->document();
+    if (!document || !document->isFullyActive() || !document->page()) {
+        promise->reject(Exception { NotAllowedError, "Document is not fully active"_s });
+        return;
+    }
+    if (!isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type::ScreenWakeLock, *document, LogFeaturePolicyFailure::Yes)) {
+        promise->reject(Exception { NotAllowedError, "'screen-wake-lock' is not allowed by Feature-Policy"_s });
+        return;
+    }
+    if (document->visibilityState() == VisibilityState::Hidden) {
+        promise->reject(Exception { NotAllowedError, "Document is hidden"_s });
+        return;
+    }
+
+    // We should be requesting permission but at the moment, we merely require a user gesture.
+    bool hasUserGesture = UserGestureIndicator::processingUserGesture(document);
```

Comment:
Should we use/consume transient activation instead?

Reply:
I updated the patch to *use* transient activation. Let me know if you think we should *consume* instead.